### PR TITLE
add System.Net.Quic assembly and use it in HttpClient

### PR DESCRIPTION
**Pull request by geoffkizer**
- Add System.Net.Quic assembly, based on existing code in shared directory
  - Note, the msquic implementation is entirely unchanged, just moved
  - The external facing classes are unchanged except for making the API public
- Add new Mock implementation (in memory only) for System.Net.Quic
- Update System.Net.Quic tests to run against both MsQuic provider (if available) and Mock provider (always available)
  - Note, a few tests haven't been updated yet
- Update System.Net.Http to use the new System.Net.Quic assembly instead of shared code
- Add SocketsHttpHandler.QuicImplementationProvider to allow specifying the QUIC implementation provider
  - This is temporary, like the whole QuicImplementationProvider infrastructure, and is not expected to ship 
- Enable HTTP3 by default in System.Net.Http if the specified QuicImplementationProvider is supported
- Update System.Net.Http tests to run against both MsQuic provider and Mock provider
  - Mock provider is always available, so these tests will now run in CI
  - MsQuic tests run when MsQuic provider is available
- Update System.Net.Http tests to disable failing or inapplicable HTTP3 tests
- Edit: The old shared QUIC code is also removed now.

@dotnet/ncl @scalablecory @stephentoub @jkotalik @halter73 @davidfowl

### CODE REVIEW

**Comment by jkotalik** [comment]
For reacting in ASP.NET Core, it looks like we'll need to wait until we get the runtime update before deleting the shared code.

Besides that, is this mostly a revert of what we had originally in 5.0, see: cfac865f9d44a443843d7dfc660a497f7d5bdf5a

**Comment by geoffkizer** [comment]
> is this mostly a revert of what we had originally in 5.0

Conceptually, yeah. The actual QUIC code is not changed here except for making the APIs public.

**Review by Tratcher — APPROVED** [review]

**Review by scalablecory — APPROVED** [review]
seems fine; did not review the MsQuic implementation as I'm assuming it's just been moved.

**Review by jkotalik — APPROVED** [review]

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs [hunk 1]
```diff
@@ -19,7 +20,7 @@ public sealed class Http3LoopbackServer : GenericLoopbackServer
 
         public override Uri Address => new Uri($"https://{_listener.ListenEndPoint}/");
 
-        public Http3LoopbackServer(GenericLoopbackOptions options = null)
+        public Http3LoopbackServer(QuicImplementationProvider quicImplememntationProvider = null, GenericLoopbackOptions options = null)
```
[the remark this answers is not in the document]
Comment (geoffkizer) [reply src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs]: thanks

**Pull request merged**